src/libstd/collections/hash/table.rs

```diff
@@ -152,51 +133,58 @@ pub fn make_hash<T: ?Sized, S>(hash_state: &S, t: &T) -> SafeHash
     // EMPTY_HASH. We can maintain our precious uniform distribution
     // of initial indexes by unconditionally setting the MSB,
     // effectively reducing 64-bits hashes to 63 bits.
-    SafeHash { hash: 0x8000_0000_0000_0000 | state.finish() }
+    unsafe { NonZero::new(0x8000_0000_0000_0000 | state.finish()) }
 }
 
-// `replace` casts a `*u64` to a `*SafeHash`. Since we statically
-// ensure that a `FullBucket` points to an index with a non-zero hash,
-// and a `SafeHash` is just a `u64` with a different name, this is
-// safe.
+// `read` casts a `*mut SafeHash` to a `*mut Option<SafeHash>`. Since we
+// statically ensure that a `FullBucket` points to an index with a non-zero
+// hash, and a `SafeHash` is Zeroable, this is safe.
 //
 // This test ensures that a `SafeHash` really IS the same size as a
-// `u64`. If you need to change the size of `SafeHash` (and
-// consequently made this test fail), `replace` needs to be
-// modified to no longer assume this.
+// `Option<SafeHash>`. If you need to change the nullability of `SafeHash`,
+// some functions need to be modified to no longer assume this.
 #[test]
-fn can_alias_safehash_as_u64() {
-    assert_eq!(size_of::<SafeHash>(), size_of::<u64>())
+fn can_alias_safehash_as_option() {
+    assert_eq!(size_of::<SafeHash>(), size_of::<Option<SafeHash>>())
```

Comment:
Alternatively, this could have a call to transmute, which (at lease at the last time I checked) fails at compile time if the two type arguments have different sizes.